2.0.0

aidantwoods opened the pull request:
Big PR to track everything for planned changes towards 2.0.0—please leave reviews here (preferably by adding a review to the relevant code site), or make PRs targeting the `2.0.x` branch if you'd like to help.

I don't plan on making any more changes here directly, rather changes can be merged into this via separate dedicated PRs.

PhrozenByte committed:
Test Parsedown against cached expect-to-pass CommonMark spec examples

This test suite runs tests the same way as `test/CommonMarkTestWeak.php`, but uses a cached set of CommonMark spec examples in `test/commonmark/`. It is executed along with Parsedown's default test suite and runs various CommonMark spec examples, which are expected to pass. If they don't pass, the Parsedown build fails. The intention of this test suite is to make sure, that previously passed CommonMark spec examples don't fail due to unwanted side-effects of code changes.

You can re-create the `test/commonmark/` directory by executing the PHPUnit group `update`. The test suite will then run `test/CommonMarkTestWeak.php` and create files with the Markdown source and the resulting HTML markup of all passed tests. The command to execute looks like the following:

    $ phpunit --group update

PhrozenByte committed:
Sync phpunit data set and CommonMark spec example numbers

PhrozenByte committed:
Add initial test/commonmark/ folder

PhrozenByte committed:
Update expect-to-pass CommonMark spec examples for f4e0234

aidantwoods committed:
Implement Emphasis

aidantwoods committed:
Implement EscapeSequence

aidantwoods committed:
Implement Link

aidantwoods committed:
Implement Image

aidantwoods committed:
Implement Markup

aidantwoods committed:
Implement SpecialCharacter

aidantwoods committed:
Implement Strikethrough

aidantwoods committed:
Implement Url

aidantwoods committed:
Implement UrlTag

aidantwoods committed:
Add required imports

aidantwoods committed:
Make Parsedown non-extendable and remove comment header

aidantwoods committed:
Remove no longer needed parts of core class

aidantwoods committed:
Rewrite to use new internals

aidantwoods committed:
Use Parsedown directly

Remove UrlsLinked adjustment--we'll have a better way
of doing that later

aidantwoods committed:
Remap text-level elements retreival

aidantwoods committed:
Adjust tests for new API
Remove tests that test old core and extension features
Comment out test for no markup independent of safe mode

aidantwoods committed:
Bump Psalm version for new static analysis features

aidantwoods committed:
Remove ignore statement for core class :)

aidantwoods committed:
Calling static functions on metatype class string is >=PHP7 only
🤷‍♂️

aidantwoods committed:
7.3 was released, so make failure blocking

aidantwoods committed:
Use latest PHP on non-unit test build staged

aidantwoods committed:
Remove debug lines

aidantwoods committed:
Test on PHP 5.5 and 5.6

Issue on pre PHP 7 may be to do with method name and not call syntax as
suspected

aidantwoods committed:
Remove some return type hints that I missed

aidantwoods committed:
More keyword related fixes for pre PHP 7

aidantwoods committed:
Fix pre PHP7 closure-as-property closure syntax

aidantwoods committed:
Fix PHP 5.6 bug

aidantwoods committed:
Constant arrays to static vars for PHP 5.5

aidantwoods committed:
Improve Link API

aidantwoods committed:
Remove meaningless interrupt check
Interrupted implies previousEmptyLines > 0 in incoming Context

aidantwoods committed:
Remove useless variables

aidantwoods committed:
This isn't used and is derivable from the indent

aidantwoods committed:
Indent offset is encompassed in Lines object

aidantwoods committed:
Test for dead code in travis

aidantwoods committed:
Remove useless line method

This will be achievable by directly restricting enabled Components

aidantwoods committed:
Make use of line name instead of lineElements

aidantwoods committed:
State management

Decouple state from final rendering
Restore initial state after parsing

aidantwoods committed:
Generalise line parsing to return Inlines before applying state

aidantwoods committed:
Require Inlines to provide a best plaintext rendering

This allows markdown to be parsed "inside" the alt
attribute of an image, and then the best plaintext
can be used as the rest.
This improves CommonMark compliance.

aidantwoods committed:
Add dead code test to composer test script

aidantwoods committed:
Thematic breaks can't have an indent of more than 3

aidantwoods committed:
Tabs shouldn't break thematic breaks

aidantwoods committed:
rtrim non hard breaking lines

aidantwoods committed:
Make it clearer this is one backslash character

aidantwoods committed:
Improve indent handling by lists

aidantwoods committed:
Better name for testing safe mode and strict mode state
Add a nice named constructor

aidantwoods committed:
Headers can't start with more than three spaces

aidantwoods committed:
Tabs are allowed after header delimiter

aidantwoods committed:
We should be using strict mode in commonmark benchmark

aidantwoods committed:
Remove closing # sequence from header

aidantwoods committed:
Don't remove right #'s too early (before dealing with whitesapce)

aidantwoods committed:
Produce Blocks before converting to StateRenderables
(As we do with Inlines)

aidantwoods committed:
Text variable isn't very useful inside the loop

aidantwoods committed:
Add hard and soft breaks

aidantwoods committed:
Allow inlines to backtrack into characters that were suspected as
inlines but ruled out
This is required for backslash escaped Hard breaks to work:
Parsedown first checks to see if these are escape sequences,
however when they are ruled out they should not be assumed to be
plaintext since a later inline (Hardbreak) may backtrack into these
unconsumed characters.

aidantwoods committed:
Default value is intrinsic to a configurable, we can just always
retrieve that.

aidantwoods committed:
Inlines and Blocks as Configurables

aidantwoods committed:
Now the class is a bit shorter we can remove these makeshift dividers

aidantwoods committed:
More keyword fixes for pre-PHP7

aidantwoods committed:
More meaningful method name

aidantwoods committed:
This should be slightly faster
Merge would honour changes, and removals
are equivalent to changing to default value.

aidantwoods committed:
Add easy way to remove Components from InlineTyes and BlockTypes

aidantwoods committed:
Put reused code in a trait so boolean configurables are easy to make

aidantwoods committed:
Add breaks configurable

aidantwoods committed:
We don't need to pass a Parsedown instance down

Since Parsedown is instancible from State, we only
need to carry that down.

aidantwoods committed:
Remove test extensions

aidantwoods committed:
Cleanup logical operators

aidantwoods committed:
Add some style cleanup rules already followed

aidantwoods committed:
Ensure Url parsing is removed all the way down (not just edge)

aidantwoods committed:
Return state after block parse instead of mutating the instance copy

aidantwoods committed:
Don't special case invisible
If something has no html, it doesn't need to have a newline

aidantwoods committed:
Caching for initial configurable values removes need to seed state

aidantwoods committed:
Fix strange formatting

aidantwoods committed:
Remove reminder comment
urlsLinked(false) is replaced by customising to remove
the inline from InlineTypes configurable when initialising
Parsedown

aidantwoods committed:
No markup can be achieved by removing the respective parsing Components

aidantwoods committed:
Don't assume marker type is correct

aidantwoods committed:
Allow empty code spans

aidantwoods committed:
Make escaping slightly less aggressive

aidantwoods committed:
CommonMark escapes double-quotes

aidantwoods committed:
Match CommonMark's rendering a bit better

aidantwoods committed:
Backslash escape like CommonMark

aidantwoods committed:
Markup like CommonMark

aidantwoods committed:
Since SafeMode concerns output, spacing should still be parsed like markup

aidantwoods committed:
As best I can tell, these were passed accidently before

aidantwoods committed:
Fix regex compilation

aidantwoods committed:
Cache new passing tests

aidantwoods committed:
Permit empty links

aidantwoods committed:
Trim paragraph contents

aidantwoods committed:
HTML tags should have a name

aidantwoods committed:
Make sure closing sequence is removed correctly

aidantwoods committed:
Track whitespace left on blank lines to match CommonMark
Test changes copy pasted to match CommonMark reference parser

aidantwoods committed:
Remove cached items no that longer exist in spec.txt

aidantwoods committed:
Fix whitespace errors

aidantwoods committed:
Switch to CommonMarkStrict tests

aidantwoods committed:
Test strict and weak in Travis

aidantwoods committed:
Require integer-keyed lists so that array_merge result is predictable

aidantwoods committed:
Add convenience instance-based initialisers

aidantwoods committed:
Add StateBearer which can carry state

aidantwoods committed:
PHP < 7 compat

aidantwoods committed:
Remove leftover hackyness

aidantwoods committed:
Substr over indexing string

aidantwoods committed:
Fix recovered spaces calculation

aidantwoods committed:
Add mutation testing

aidantwoods committed:
Nightly doesn't have xdebug

aidantwoods committed:
Run mutation tests after unit tests on supported platforms

aidantwoods committed:
Length not necessary

aidantwoods committed:
Test continuation of indented code blocks

aidantwoods committed:
Test against HTML block endings

aidantwoods committed:
Ensure markers are checked when beginning blocks

aidantwoods committed:
Paragraph would end itself and new one must start for header row to
exist

aidantwoods committed:
--show-mutations over cat infection.log

aidantwoods committed:
Cache spec locally for 5 minutes

aidantwoods committed:
PHP 5.5 compat

aidantwoods committed:
Simplify expression

aidantwoods committed:
Test trimming doesn't occur when asymmetric

aidantwoods committed:
Fix whitespace trimming for soft and hard breaks

aidantwoods committed:
Remove deleted stage

aidantwoods committed:
Test blockquote whitespace handling

aidantwoods committed:
Acquisition capable blocks as an interface

aidantwoods committed:
Move url sanitisation out of Element class

aidantwoods committed:
Remove redundant checks

These don't appear to have a measurable positive impact on performance.

aidantwoods committed:
Missed one

aidantwoods committed:
Tighten requirements

aidantwoods committed:
Require State to build and advance blocks

aidantwoods committed:
Collect State from continuable state updating block on advance

aidantwoods committed:
Assert marker is correct for hard and soft breaks

aidantwoods committed:
Early exit if found

aidantwoods committed:
Update excludes

aidantwoods committed:
Ensure $startPosition is positive

aidantwoods committed:
Type check tests

aidantwoods committed:
Ensure marker is properly contained in the Inline

aidantwoods committed:
These are self-explainitory

aidantwoods committed:
Remove inaccurate comment

aidantwoods committed:
Remove unused import

aidantwoods committed:
Remove more redundant checks

aidantwoods committed:
Simplify

aidantwoods committed:
Suppress instead of writing super verbose tests

aidantwoods committed:
Remove unused methods

aidantwoods committed:
Add some component level tests

aidantwoods committed:
These classes of mutations would be caught be static analysis
Infection doesn't support running Psalm, so removing
the mutations is next best solution.

aidantwoods committed:
Improve tests

aidantwoods committed:
Further improve tests

aidantwoods committed:
No contexts implies everything being appended is a blank line
(of which where is at least one).

aidantwoods committed:
Test indented code boundary when list advances

aidantwoods committed:
Make this pattern a bit more reusable

aidantwoods committed:
Constructor shouldn't be part of public API

aidantwoods committed:
Expand public API of Components
Ref: https://github.com/erusev/parsedown/issues/694

aidantwoods committed:
Trim in renderer

aidantwoods committed:
Remove copy-pasto

aidantwoods committed:
Backtracking capable inlines better expressed by interface

aidantwoods committed:
Ensure we cover all mutations in tests

aidantwoods committed:
Normalise link reference lookups
Ref: https://github.com/erusev/parsedown/pull/104

aidantwoods committed:
Use caret operator for PHP version
As suggested by @glensc

aidantwoods committed:
Update README to match Travis breadth

aidantwoods committed:
Match GFM spec for extended autolinks
There is a small list of characters which may precede an autolink

aidantwoods committed:
According to GFM spec these shouldn't infact be recognised

aidantwoods committed:
Update CommonMark expect to pass

aidantwoods committed:
Remove format error

aidantwoods committed:
Port fix for https://github.com/erusev/parsedown/issues/699 into new
codebase

aidantwoods committed:
PHP < 7 patch

aidantwoods committed:
Update commonmark cache

aidantwoods committed:
Limit recursion depth by configurable
Fixes https://github.com/erusev/parsedown/issues/681

aidantwoods committed:
Update commonmark cache

aidantwoods committed:
More specific return type

aidantwoods committed:
Fix malformed return type

aidantwoods committed:
Allow unused class in tests

aidantwoods committed:
Update test cache

aidantwoods committed:
Drop pre 7.1
Drop pre 5.6 for constant arrays
Drop pre 7.1 for private constants

aidantwoods committed:
Add type 7 HTML block tests

aidantwoods committed:
Account for type 7 HTML blocks
These differ from type 6 blocks in a few ways (other than the HTML
element names given in the spec).

1. Type 7 blocks cannot interrupt paragraphs
2. Type 7 block must contain ONLY whitespace on the rest of the
   line, whereas type 6 blocks need only be followed by whitespace
   after the tag name, or have the opening tag be complete.

aidantwoods committed:
Update expect-to-pass cache

aidantwoods committed:
Integer keys aren't necessary

aidantwoods committed:
Utilise constant arrays over static vars

aidantwoods committed:
Tables always acquire previous

aidantwoods committed:
Drop use of undefined reference

aidantwoods committed:
Force return type to T

It's currently not a planned feature to allow Psalm to understand
the type relationship of class-name => class-instance as the array
here is constructed.
Ref: https://github.com/vimeo/psalm/issues/1969

aidantwoods committed:
Bump min psalm version

aidantwoods committed:
Prefer master until next release

aidantwoods committed:
Drop master from psalm requirement

aidantwoods committed:
$matches is guarenteed to contain data for the full pattern

aidantwoods committed:
More specific annotation

aidantwoods committed:
Formatting fix

aidantwoods committed:
Add 7.4 snapshot

aidantwoods committed:
Only run units on 7.4snapshot

nathanlesage commented:
Hey, thanks for the heads up over at the october-repo. I understand that you need reviews and honestly, I haven't looked at the PR because I don't have such a great knowledge of this repository's source, but if it helps, I'll see to managing a review around Christmas, even a short one.

aidantwoods commented:
Thanks so much @nathanlesage. Even a review from the prospective of a user of the API would be helpful, sans a full on code review which I realise would be quite a substantial task. 

There's sort of a summary of the motivation for changes [in a comment here](https://github.com/erusev/parsedown/issues/601#issuecomment-452206070) and a [second one overviewing some changes here](https://github.com/erusev/parsedown/issues/601#issuecomment-455895576), both these from the same thread which discussed a potential feature release 1.8.

A big problem I'm looking to solve is to draw a clear API boundary for the equivalent of "extensions" to utilise in the context of 2.0. Unfortunately a lot of extensions to Parsedown 1.x ended up depending on `protected` behaviour, which meant it was very hard to guarantee new features (or even bug fixes in some cases) were non-breaking even with [best attempts](https://github.com/erusev/parsedown/issues/601#issuecomment-387080763) at adding compatibility layers where internal behaviour had changed. This was made especially problematic because this way of doing things was essentially per-instructions when looking at the [Wiki](https://github.com/erusev/parsedown/wiki/Tutorial:-Create-Extensions#change-element-markup) on making extensions.

There are also a lot of bugfixes either pulled from that 1.8 thread, or found while rewriting the library and referencing spec, adding more tests, or stumbled upon by mutation testing (e.g. [I added a test to catch a specific mutation and ended up finding a bug](https://github.com/erusev/parsedown/pull/685/commits/2049587518c75b8e2dd6372529d9bf750832072c)).

Also some feature additions I'm quite pleased with, e.g. [a recursion limiter](https://github.com/erusev/parsedown/pull/708/commits/9eb6a02334a4728e13971163db5f6b6aae7be833) which I was able to add using a `Configurable`, which is how I'd envision others add custom behaviour like setters or preserved parsing state. Or another example: [`DefinitionBook`](https://github.com/erusev/parsedown/blob/2.0.x/src/Configurables/DefinitionBook.php) is a `Configurable` which [inline `Link`s can utilise](https://github.com/erusev/parsedown/blob/0a6408043f052f989a2be1c6dd75edefadbe18b2/src/Components/Inlines/Link.php#L84) to lookup link definition data (this was previously an array stored in the core Parsedown class).

nathanlesage commented:
Heya,

so I just spent half of the day for an intensive code review, and I'm pleased to tell you that I'm through with it! Please remember that this is my **very first code review**, so if I did anything wrong or if something is still missing, please tell me! 🙈 Furthermore, I'm not completely familiar with the codebase and conventions you're using, so please, if something I remark contravenes these, just ignore that! :)

Most changes concern semantic sugar for naming conventions, so it ultimately boils down to taste. Some were somewhat bigger, so first a quick rundown:

## In General

- I really like how the code is built conceptually, with a good AST and clean, separated functionality.
- However, some naming conventions, especially for some setters and getters are a little bit confusing. In general, I feel like adding certain expressive keywords such as "has", "get", and "set" to the beginning of certain functions would help readability of the code. But I'm coming from a JavaScript/NodeJS background, where this is extremely common.
- Concerning function naming conventions, I think that using continuous forms of verbs (appending instead of append) is difficult, because after all, this is very uncommon and if you call a function, it will complete running before any other expression runs; I've included specific examples below!
- All Block and Inline elements reference the StateRenderable class which is neither accessed nor otherwise implemented by these, as the class is already being used upstream in the Component class, which these inherit. While I am in no way an authority on how PHP loads classes, I get the impression that we could remove these references, plus some more (I could imagine that some of the classes in the top simply were forgotten during the extensive refactoring). Plus, I spotted some other lone remnant classes that were not being used by the classes I inspected.
- While I understand now your idea of implementing the RecursionLimiter as a Configurable, I have my doubts that this should be like this. Personally, if I understand the RecursionLimiter correctly, it's a bail-out option for the parser to not enter infinite loops, so I feel as if this should be a core class (because in the end its functionality also supersedes the one of other Configurables, it has a really huge priority in the call stack). But I mean as you explicitly stated that you wanted to implement it as a Configurable, I'm not persisting on that issue :)
- Additionally, I saw that a lot of classes' functions do not manipulate themselves, but actually re-create themselves with the new option (especially in the State-class, see below for examples). I always have the feeling that this might increase computational burdens because it basically re-parses everything that's already correct, and I think that having two lines (one to apply the change, and one to `return $this`) is to be preferred over re-instantiating the same object over and over again. Or do I miss the elephant in the room with this one?

Furthermore, I ran the PHPUnit and it said that approx. 192 tests were failing, but if I remember correctly, at some point I saw someone saying that some specific tests still fail, which is fine, right? Just wanted to drop that here!

Now to some internals -- first and foremost, I did not closely inspect EVERYTHING, beause after a while, many of my remarks were recurring (especially naming conventions), so here now all my notes I took during the review, unfiltered:

***

### In file ./Parsedown.php:

- text()
    - I would recommend to add an Alias "renderMarkdown" for semantic clarity. Calling "Parsedown->text()" is a little bit less expressive. In the end, what `text()` does is run a full cycle from the input to the final HTML output. Maybe even `toHTML` as an Alias …? (This is only something I personally wanted to have with Parsedown for years now, so it's not that important, especially given the fact that we need the text-method to persist for backwards-compatibility :D)

### In file ./State.php:

- state()
    - Recommend removing this function, both because it _is_ a State object and to expect $state->state() to only return $state seems a little bit counterintuitive. Furthermore, what does one need a StateBearer-class for, other than encapsulating a State-object? I assume, by removing the state()-function one would need to also remove that class, right?
- setting()
    - This function name also sounds a little bit inexpressive. If I understand the function correctly, it adds one Configurable to the State, so wouldn't it be much easier to understand if you call State->addConfigurable($myConf)?
    - Additionally, wouldn't it be computationally easier to simply array_merge the new configurable to the already existing state array ...?
- mergingWith()
    - I would strongly recommend to rename it to "mergeWith" (this pattern also exists in other files, and it sounds really weird to me).
    - Furthermore, as well as the setting()-function: Wouldn't it be easier to simply array_merge the two states, as they after all are only key-value-arrays of Configurables?

### In file ./Parsing/Lines.php:

- __construct()
    - Same as with the Context::__construct() -- is it necessary to both save the trailing line break characters _and_ their count? I see that this will enable Parsedown to fully reconstruct the input Markdown, but as whitespace on a line has no meaning in the context of Markdown, my neurotic self somehow wants this gone...
- fromTextLines()
    - The variable $sequentialLines should be renamed to $preceedingEmptyLines (or similar) for clarity
- none()
    - Maybe rename to createEmpty() for clarification?
- isEmpty()
    - Wouldn't it be computation-wise better to check for the boolean var "containsBlankLines"?
- Contexts()
    - Maybe rename to "getContexts"? (But I'm really not sure if getters/setters are semantically used throughout the code! I just stumbled upon the uppercase first letter.)
- trailingBlankLines()
    - Same (with "has") ...?
- appendingBlankLines()
    - Strongly recommend to rename it to "append" instead of "appending".
    - Signature mismatch: It does not return $this, but a new, cloned Lines-object.
    - Line 136: It assumes that the newly created $NextLines object does in fact contain blank lines, but I would recommend to copy the $NextLines->containsBlankLines-property over.
    - Also, I have the question why the appending-functions do not append the lines to themselves, and instead return a new Lines-object ...? On a practical level, the only time I would like to append the lines to the already existing object; or is there a use-case I'm missing here?
- appendingTextLines()
    - Strongly recommend to rename it to "append" instead of "appending".
- appendingContext()
    - Strongly recommend to rename it to "append" instead of "appending".
    - Recommend to simplify the if-statement to set the containsBlankLines with an inline-expression
    - The resetting of the trailing blank lines seems a little bit counterintuitive to me, as this does not happen when you append TextLines (which, strictly speaking, are the same with a little bit less information), so if I append TextLines, the trailing blank lines are preserved, while when I add a context, they are omitted.

### In file ./Parsing/Context.php:

- __construct()
    - Is it really necessary to also save the $previousEmptyLinesText? As far as I am concerned, it seems that it can only contain linebreaks, which are more efficiently saved using the count variable.
    - Maybe exchange the "previous" in the variable and function names with "preceeding" (or semantically similar), because the functions and variables will only be accessed in context of a Context object, so $myContext->preceedingEmptyLines makes more sense then $myContext->previousEmptyLines (the latter might refer to _all_ previous blank lines, even those from former contexts). But I agree this is maybe borderline hair splitting on my part!

### In file ./Configurables/RecursionLimiter.php

- increment()
    - I've mentioned this before, but is the benefit of only having one line worth it to always re-create objects in memory, or wouldn't it suffice to actually only increment the pointer?

***

Finally, I tried to set up a fresh install of OctoberCMS, but since Catalina, it won't install on my machine anymore, so I couldn't really try it out :( Also I didn't take notes on all of the classes, but I've had a thorough look at them, and could not spot anything else than other instances of what I've just mentioned with these core classes!

I hope that this information helps, and I would like to apologise for the wall of text!

BenjaminHoegh commented:
I have to agree to multiple of @nathanlesag and I strongly agree with changing `text()` to `toHTML()` in Parsedown.php

And a little wish from me, if you make these changes would you please write back, so I can begin the development of my plugins to work with 2.0 when you release it

aidantwoods committed:
Simplify match logic

aidantwoods committed:
Use list type over int array

aidantwoods committed:
Use ->toHtml over ->text

aidantwoods committed:
Compute on read optimisation for previousEmptyLines

aidantwoods committed:
Clarify sequential lines being tracked are empty

aidantwoods committed:
Expensive operation last

aidantwoods committed:
Fix uppercase typo

aidantwoods committed:
previousEmptyLines -> precedingEmptyLines

aidantwoods commented:
Hi @nathanlesage,

Apologies for not getting to this sooner—time's been a bit scarce recently and I wanted to be able to give your review a proper run though. Thank you so much for taking the time, and going to the level of detail you have!

I'll structure my reply similarly to how you've left your review, to make navigation of the individual points I'm replying to a bit easier :)

## In General

> * Concerning function naming conventions, I think that using continuous forms of verbs (appending instead of append) is difficult, because after all, this is very uncommon and if you call a function, it will complete running before any other expression runs; I've included specific examples below!

This is something I've picked up from Swift, so is sort of intentional:

https://swift.org/documentation/api-design-guidelines/#name-according-to-side-effects

> #### Name functions and methods according to their side-effects
> [...]
> When the operation is naturally described by a verb, use the verb’s imperative for the mutating method and apply the “ed” or “ing” suffix to name its nonmutating counterpart.
> Mutating      | Nonmutating
> ------------- | ------------------
> x.sort()	    | z = x.sorted()
> x.append(y)	| z = x.appending(y)

The idea here is to form a distinction between whether a change occurs in-place (mutating) or whether it occurs only in the returned result of the function being called (non-mutating). Most things in the 2.0 branch are immutable (I think I've made an exception for result caching), and so perhaps this distinction could be dropped. I do quite like the idea of giving a hint as to the side-effects of a method through its naming though (especially absent any language-level aids for gauging this).

> * However, some naming conventions, especially for some setters and getters are a little bit confusing. In general, I feel like adding certain expressive keywords such as "has", "get", and "set" to the beginning of certain functions would help readability of the code. But I'm coming from a JavaScript/NodeJS background, where this is extremely common.

There shouldn't be any (mutating) setter functions in the codebase, and so I think we'd end up in a situation where almost every property associated method is prefixed with "get"? Perhaps I'm missing something here though, is there some ambiguity about what `$Image->url()` would be expected to do over `$Image->getUrl()`? In the situation where there are arguments being passed I would tend to agree with adding more detail, but when there are no arguments and the function corresponds directly to the property name, is this actually clearer?
In an idea world, properties could be marked as non-mutable and it would be possible to make them public directly without throwing side-effects to the wind :)

Did you find any examples missing a "has"? I'd definitely be interested in fixing those!

> * All Block and Inline elements reference the StateRenderable class which is neither accessed nor otherwise implemented by these, as the class is already being used upstream in the Component class, which these inherit. While I am in no way an authority on how PHP loads classes, I get the impression that we could remove these references, plus some more (I could imagine that some of the classes in the top simply were forgotten during the extensive refactoring). Plus, I spotted some other lone remnant classes that were not being used by the classes I inspected.

I think you're referring to the `use` statement at the top of the files? You're correct here, I have a fixer running meant to remove stuff like this but I think it has avoided removing that line due to the function `stateRenderable` sharing the name :)

That said, I'm thinking it might be a good idea to push the return types back up to being `StateRenderable`, rather than the specific thing being returned. The idea here is that this allows the return type to be to be changed without breaking the API (due to an unnecessarily specific return type being declared).

> * While I understand now your idea of implementing the RecursionLimiter as a Configurable, I have my doubts that this should be like this. Personally, if I understand the RecursionLimiter correctly, it's a bail-out option for the parser to not enter infinite loops, so I feel as if this should be a core class (because in the end its functionality also supersedes the one of other Configurables, it has a really huge priority in the call stack). But I mean as you explicitly stated that you wanted to implement it as a Configurable, I'm not persisting on that issue :)

The idea with implementing it as a configurable is that it uses the "condoned" method of preserving parsing state. Rather than accessing special "magic" properties in the core class (or worse, global ones), everything now has the ability to add and use configurables as they please (and in a way where mutation of those configurables has a clear scope).

> * Additionally, I saw that a lot of classes' functions do not manipulate themselves, but actually re-create themselves with the new option (especially in the State-class, see below for examples). I always have the feeling that this might increase computational burdens because it basically re-parses everything that's already correct, and I think that having two lines (one to apply the change, and one to `return $this`) is to be preferred over re-instantiating the same object over and over again. Or do I miss the elephant in the room with this one?

The elephant in the room here is mutability (and localisation of it) :) By returning a new version, the caller always has a clear copy of both the original and the modified copy (with no unexpected surprises). If the objects where to mutate themselves, it's not always obvious when a change has occurred (and to the extent of the scope). Without immutable objects, and because objects are passed around quite a lot, there might be situations where context A passes an object to context B, which modifies it for its own working purposes. But because objects in PHP are passed by handle, this also modifies context A's copy which might result in unexpected behaviour.

> Furthermore, I ran the PHPUnit and it said that approx. 192 tests were failing, but if I remember correctly, at some point I saw someone saying that some specific tests still fail, which is fine, right? Just wanted to drop that here!

Hm, that is odd actually—I'm not getting any failing tests. There is a mode where we run against all CommonMark tests (of which some fail due to not being fully compliant), we then cache and commit the passing ones to prevent regression. Are you using the version of phpunit download by the project's composer settings (`vendor/bin/phpunit`), or are you using a different one?

## In file ./Parsedown.php:
> * I would recommend to add an Alias "renderMarkdown" for semantic clarity. Calling "Parsedown->text()" is a little bit less expressive. In the end, what text() does is run a full cycle from the input to the final HTML output. Maybe even `toHTML` as an Alias …? (This is only something I personally wanted to have with Parsedown for years now, so it's not that important, especially given the fact that we need the text-method to persist for backwards-compatibility :D)

(a72455c) I like this idea, `text` is a bit unclear. I think I prefer `toHTML` (or maybe `toHtml`?) to `renderMarkdown` (maybe I'm just being a lazy typist here :) ).

## In file ./State.php:
> * state()
>   * Recommend removing this function, both because it is a State object and to expect $state->state() to only return $state seems a little bit counterintuitive. Furthermore, what does one need a StateBearer-class for, other than encapsulating a State-object? I assume, by removing the state()-function one would need to also remove that class, right?

This is due to the `StateBearer` interface (I might rename this). This would be the interface that I'd envision today's "extensions" implementing, in order to arrive at a sort of packaged initial state for parsing (which contains, among other things, which type of parsing will be done). Because there's an obvious way for `State` to directly implement this (return itself) I thought it might be nice to be able to pass that directly into Parsedown's constructor too (instead of being forced to create a wrapper class).

> * setting()
>   * This function name also sounds a little bit inexpressive. If I understand the function correctly, it adds one Configurable to the State, so wouldn't it be much easier to understand if you call State->addConfigurable($myConf)?

Maybe `addConfigurable` might be a better name here (though it doesn't *quite* form an additive step). `State` is essentially a localised singleton store. So while the objects it store's aren't singletons, there is guaranteed to be one (and only one) of each one in the store. In this sense it sort of "sets" rather than "adds", but we can probably do better than `setting`.

>   * Additionally, wouldn't it be computationally easier to simply array_merge the new configurable to the already existing state array ...?

Computationally easier yes, but (especially with this object) a significant loss to stability. State being immutable allows its mutation to be localised without deep-cloning (and mutations that don't propagate everywhere unpredictably are quite important in a few areas of the code).

e.g. the removal of autolinking Urls from subparsing (and only subparsing) after parsing a link https://github.com/erusev/parsedown/blob/0a6408043f052f989a2be1c6dd75edefadbe18b2/src/Components/Inlines/Link.php#L135-L137

## In file ./Parsing/Lines.php:
> * __construct()
>   * Same as with the Context::__construct() -- is it necessary to both save the trailing line break characters and their count? I see that this will enable Parsedown to fully reconstruct the input Markdown, but as whitespace on a line has no meaning in the context of Markdown, my neurotic self somehow wants this gone...

Unfortunately hanging onto the actual break characters is necessary. Computing the count in the constructor is done for the sake of avoiding the recompilation if the count is called multiple times (though this can probably be optimised to be a cached result after first read in `Context` (8d09320), in `Lines` this value is already used in the constructor). When outputting fenced (or indented) code blocks the exact characters used on "empty" (whitespace only) lines ends up being important for CommonMark purposes.

> * fromTextLines()
>   * The variable $sequentialLines should be renamed to $preceedingEmptyLines (or similar) for clarity

`$preceedingEmptyLines` makes sense for the body of the loop, but its final value is actually the number of *trailing* empty lines (when it is used to construct the final `Lines` object). `$sequentialLines` is sort of an accumulator of observed empty lines at the current point in the parse, though `$sequentialEmptyLines` is probably a better name than `$sequentialLines` (39b8b04).

> * none()
>   * Maybe rename to createEmpty() for clarification?

We can probably improve on `none` :) I'm wondering if `createEmpty` might give the impression of creating an empty line though? The objective here is for zero lines to be present.

> * isEmpty()
>   * Wouldn't it be computation-wise better to check for the boolean var "containsBlankLines"?

Would this be opposed to the `&& $this->trailingBlankLines === 0` part of the check? I suppose this could be marginally faster than the comparison to zero? At the very least I could reorder that to be first too, so count is short-circuited in the case that blank lines are present (525b723).

> * Contexts()
>   * Maybe rename to "getContexts"? (But I'm really not sure if getters/setters are semantically used throughout the code! I just stumbled upon the uppercase first letter.)

(7610eac) Upper first letter is a typo here :) If we decide to use `get` prefix throughout for getters then I'll add that as part of a bulk refactor.

> * trailingBlankLines()
>   * Same (with "has") ...?

The method returns the *number* of blank lines, though it looks like this isn't used publicly anywhere anymore (so can possibly remove this method).

> * appendingBlankLines()
>   * Strongly recommend to rename it to "append" instead of "appending".
>   * Signature mismatch: It does not return $this, but a new, cloned Lines-object.

The return type here is `self`, not `$this`?

>   * Line 136: It assumes that the newly created $NextLines object does in fact contain blank lines, but I would recommend to copy the $NextLines->containsBlankLines-property over.

Because `\count($NextLines->Contexts) === 0`, the `$NextLines` must contain at least one blank line because it was built from text (shortest example is that it was the empty string), and so `$Lines->containsBlankLines = $Lines->containsBlankLines || $NextLines->containsBlankLines` is always equivalent to `$Lines->containsBlankLines` because `$NextLines->containsBlankLines === true`. It used to be similar to this, but was changed due to the surrounding `if` statement: https://github.com/erusev/parsedown/commit/a396fccace3dddf06fb21561566700c279162d09 (actually found this one through mutation testing, and was initially a bit confused when I saw the assignment could be [mutated](https://travis-ci.org/erusev/parsedown/jobs/491826152#L1484-L1496) without detection by the test suite).

>   * Also, I have the question why the appending-functions do not append the lines to themselves, and instead return a new Lines-object ...? On a practical level, the only time I would like to append the lines to the already existing object; or is there a use-case I'm missing here?

Again this is to do with mutability. To give an example, Parsedown itself uses a `Lines` object. If Parsedown were to hand this down to a particular block for parsing, and the block appended some lines to it for its own working purposes, Parsedown's copy would also be modified. For this reason immutable objects are used everywhere, so that copies only need to be made when changes are made (and not every time you pass an object to a new context). For this reason mutations (through redefining of a variable) are localised to small function/object scoped contexts, rather than propagating in unexpected ways.

> * appendingContext()
>   * Strongly recommend to rename it to "append" instead of "appending".
>   * Recommend to simplify the if-statement to set the containsBlankLines with an inline-expression

This is subtle, but not equivalent to the condition in the `if` on its own. If `$Lines` already contains blank lines, but the `Context` being appended doesn't have any preceding blank lines, then an inline statement `$Lines->containsBlankLines = ($Context->previousEmptyLines() > 0)` would change the `true` to `false`, even though no blank lines were removed.

Best we could do with an inline expression would be `$Lines->containsBlankLines = ($Lines->containsBlankLines || $Context->previousEmptyLines() > 0)`, which is quite a long line. Do you think it is still worth inlining at this stage? Potentially it is actually slightly clearer because it states both criteria explicitly?

>   * The resetting of the trailing blank lines seems a little bit counterintuitive to me, as this does not happen when you append TextLines (which, strictly speaking, are the same with a little bit less information), so if I append TextLines, the trailing blank lines are preserved, while when I add a context, they are omitted.

When you append text lines, these might contain multiple `Context`s (a `Context` is a single non-whitespace only line, paired with any whitespace only lines that occurred before it). For this reason, when appending text lines, a new `Lines` object is used to parse this block of text, which also parses any associated trailing blank lines at the end ([this information is copied across when merging with the current `Lines`](https://github.com/erusev/parsedown/blob/0a6408043f052f989a2be1c6dd75edefadbe18b2/src/Parsing/Lines.php#L148-L149)).

When you append a `Context` it, by definition, doesn't have any information about trailing lines. For this reason, any trailing lines that were in place are prepended to the lines in the `Context` being added, but because the `Context` can't itself contain information about any lines after itself (and it is now the last `Context`), there are no blank lines at the end (hence the reset).

## In file ./Parsing/Context.php:

> * __construct()
>   * Is it really necessary to also save the $previousEmptyLinesText? As far as I am concerned, it seems that it can only contain line breaks, which are more efficiently saved using the count variable.

Yes, if some `Context`s are within a code block, it is important to preserve the exact text which was present. Following on from your previous suggestion, I've delayed (8d09320) actually computing the count until it is actually used (at which point it is cached).

>   * Maybe exchange the "previous" in the variable and function names with "preceding" (or semantically similar), because the functions and variables will only be accessed in context of a Context object, so $myContext->preceedingEmptyLines makes more sense then $myContext->previousEmptyLines (the latter might refer to all previous blank lines, even those from former contexts). But I agree this is maybe borderline hair splitting on my part!

This is a nice semantic clarification (a6c17f4) :) This aligns closer with the "trailingBlankLines" wording used in `Lines` too (i.e. it doesn't read "nextEmptyLines").

## In file ./Configurables/RecursionLimiter.php
> * increment()
>   * I've mentioned this before, but is the benefit of only having one line worth it to always re-create objects in memory, or wouldn't it suffice to actually only increment the pointer?

This is actually a great example of where immutability saves us from an unexpected failure :) If I make the following change:

```diff
  /** @return self */
  public function increment()
  {
+     $this->currentDepth++;
+     return $this;
-     // return new self($this->maxDepth, $this->currentDepth + 1);
  }
```

then we end up with 543 test failures :p

This is because we end up sharing a single recursion limiter across different parsing depths, and the recursion limiter is incremented each time we increase the depth.

Imagine we have two blocks, each with their own content that needs parsing. After we parse the blocks (parse level 0), we increment the recursion limiter when we enter parsing of the contents of the block. When we parse the contents of the second block, we also increment the recursion limiter. Because we're referring to the same recursion limiter each time, we end up incrementing it twice, even through we only went one level deep. If we had 100 blocks, this very quickly diverges from the value it should be. By having an immutable recursion limiter, we only increment the copy that we're passing down to "deeper" parsing layers, and so the value is incremented once (and only once) per parsing layer (as expected).

> [...]I didn't take notes on all of the classes, but I've had a thorough look at them, and could not spot anything else than other instances of what I've just mentioned with these core classes!
> 
> I hope that this information helps, and I would like to apologise for the wall of text!

Thank you again for taking the time to look through! I've made some commits (noted inline) to address what I can immediately from your feedback. Please feel free to argue with any of my responses :) In some cases I've tried to clarify the reasoning for why something has been done the way it has, rather than making a change straight way. If you still think a change is worth making in these cases, we can definitely look at doing that though! :)

aidantwoods committed:
Use verb imperative to indicate method is non-mutating

aidantwoods committed:
PHP 7.4 is now released

nathanlesage commented:
Hey, whoa thanks for the extensive answer! :D Here's an also-long answer, but I tried to boil it down :)

(TLDR at the end)

> The elephant in the room here is mutability (and localisation of it)

TADAAAAAA 🎉

In fact, I think that _a lot_ of my complaints have indeed to do with mutation/immutability. I have to confess that I'm extremely cautious of having more than one copy of a certain big object (such as a full parsing tree) in memory, but I think this comes both from my times as a young child when we had like 8 megabytes of RAM, and all literature on programming I read told me "OMG DON'T ADD THIS VAR IF YOU DON'T NEED TO!!" which somehow burned itself into my head. Also, in JavaScript contexts, immutability is close to impossible so one doesn't even bother to try :D (If you want to see how to return minimal object copies, [here you go](https://github.com/Zettlr/Zettlr/blob/4a608194ac929101e0e913da6df3a24292234d83/source/main/zettlr-file.js#L590-L609))

This explains a lot of the stuff, and so even the `appending`-form makes sense. Albeit, I'd argue that "appended" (or even better "withAppended," so that you'd arrive at something like "withAppendedContext" to hurt your keyboard extra, hehehe) would be better (because then you have the semantic clarity of immutability without implying the function is always running), but again: borderline hair splitting :D 

Concerning the … \*looks inside the original post\* …  ah, assumption of lines in the Lines.php-class … if I understand [this line correctly](https://github.com/erusev/parsedown/blob/bded7a6ff34f9f0f31be2c60b98e973508e3bf27/src/Parsing/Lines.php#L136), wouldn't it make more sense to simply check if `$Lines->trailingBlankLines > 0`…? I am sure this breaks things, so please don't take my word for it, I am just also insecure concerning assumptions!

> Do you think it is still worth inlining at this stage? Potentially it is actually slightly clearer because it states both criteria explicitly?

Well, do you really need `$Lines->containsBlankLines = ($Lines->containsBlankLines || $Context->previousEmptyLines() > 0)`? … aaaah nononono forget that. You're right, I just realised that. It's late :D

> then we end up with 543 test failures :p

… well, it would not consist of simply leaving the RecursionLimiter that way, but to decrease the limit again once you return to level 0 :P You would need to add a decrement each time, hehehe

My only concern was that if you have multiple copied state objects, then you might end up with something like 2048 nested objects, and I don't really know if this really is necessary…? I only think that if you reach more than 256 nested elements in Markdown, you both deserve a prize and should do the walk of shame, right? 🤷🏼‍♂️ But again, this has to do with my tendency to double- and triple-check **everything** because I can't trust any computer or user …

## TLDR

In the end, what caused 99 percent of my confusion was immutability, what I completely did not have in mind — here you can see how much programming paradigms (and even single programming languages with their specific behaviours) can influence your thinking!

I trust you with all of the stuff, because once I mentioned it you had a second look at the (potential) issues, and if it still looked good for you, this means that it is really good! I really appreciate all your hard work, because if I understand it correctly, you've done the lion's share of the work yourself, and I think you should be proud of what you've done here. I know myself how hard it is to do something solely in your free time, so please, if **you** are happy with what you have done, proceed :) My aim was to pose questions (most of which were not really valid, because of the immutability-fallacy I fell for) just to have a second look, and as I already mentioned: I did not spot any obvious mistakes in the logic, and I think if you think it's ready to go out, then go for it! :)

Hope all of that helps, and thanks again! Have a good night!

P.S.: Concerning the PHPUnit: Yes, I simply ran a `composer install`, so it should've installed the correct version; but as I never used PHPUnit before (I like to live dangerously, or, in other words: for long I had no idea what testing was good for …), it might very well be an error on my side! I did run all tests, including the CommonMarks, this is why I was mentioning it!

P.P.S.: Yes, `toHtml` is also good, as long as there is such an alias :)

aidantwoods committed:
Add flag to get type-coverage

nathanlesage commented:
Hey @aidantwoods — is there anything new on this issue? :)

aidantwoods commented:
@erusev Any thoughts on the direction of the changes in this PR? I'd like to bring a branch of Parsedown-Extra in line with this codebase, get this ready for merge soon—this has probably been left alone long enough :)

The general idea of these changes is to compartmentalise functionality, allow the source code the be analysed for type errors (using [`vimeo/pslam`](https://github.com/vimeo/psalm)), and to formalise a lot of functionality that is loosely "supported" for extensions currently, but not easy to work with (especially when changing internal behaviour, which currently carries a high risk of breaking extensions).

To avoid breakage, extensions are now encouraged to utilise a defined API for accessing parsed metadata, and do not need to rely on extracting information from the exact HTML structure being generated (which is subject to change at any time—e.g. to fix bugs).

The changes here also bring some additional features not before possible, for example working with multiple extensions which do not need to know of one-another's existence, user composition of individual elements from many extensions, and the ability for these many extensions to define their own user configurable options or preserve parsing state.

No major changes have been made to the core parsing strategy (the general approach to building blocks and inlines), changes are focused on formalising previous concepts and allowing them to be more readily reusable.

erusev commented:
@aidantwoods Looks good! Thanks for the thorough explanation!

aidantwoods committed:
Add HeaderSlug configurable

Adds HeaderSlug configurable, with the option for the slug function
to be customised.

Co-authored-by: netniV <netniv@hotmail.com>

aidantwoods committed:
Remove 'matrix' key

aidantwoods committed:
Commit lock file so static analysis results are stable

aidantwoods committed:
Strip superscripts and divisions from eventual slug

As suggested by @Ayesh

Co-authored-by: Ayesh Karunaratne <ayesh@ayesh.me>

aidantwoods committed:
Resolve dependencies when running unit tests

aidantwoods committed:
Add MutableConfigurable
Hopefully I do not regret this...

aidantwoods committed:
Implement DefinitionBook as MutableConfigurable
This is a slightly more correct implementation, but perhaps not worth
the headache.

aidantwoods committed:
Trim leading and trailing hyphens from slug

aidantwoods committed:
Add SlugRegister so IDs are not duplicated

aidantwoods committed:
Merge pull request #768 from aidantwoods/feature/mutable-configurable

Add mutable configurable

aidantwoods committed:
Merge pull request #767 from aidantwoods/enhancement/header-slug

Create ID's for Header elements so they can be referenced in anchor tags

BenjaminHoegh commented:
Just out of interest, how long do you estimate you are with version 2.0. Is it too soon to begin the development of extensions?

aidantwoods commented:
@BenjaminHoegh I've let this linger quite a while, haven't I... 😄 

By all means feel free to work on an extension based on these changes. Any feedback you have would actually be really helpful (e.g. if you run into some issue with the new codebase that prevents you doing what you'd like, would be really good to know that so I can make changes if necessary).

aidantwoods commented:
Just to track it here, if memory serves I think I had the following on my list of things to do before this could be released:

- [ ] Update Parsedown Extra to work with this codebase
- [ ] Write some documentation about working with the new codebase for extension authors (especially with regard to the `State` object, and making custom "configurables"

aidantwoods committed:
Ignore phpunit result cache

aidantwoods committed:
Update dependencies

aidantwoods committed:
Fix psalm errors

aidantwoods committed:
Test 8.0 in travis

aidantwoods committed:
Better autolinks

This doesn't follow gfm spec yet, work lifted from my code over
at
https://github.com/Parsemd/Parsemd-PHP/blob/8505e2737e43f89c09fc9597889e56a6f8e114b9/src/Parsers/GitHubFlavor/Inlines/AutoLink.php

Fixes #505
Fixes #717

aidantwoods committed:
GitHub Actions

aidantwoods committed:
Remove infection for unit tests

aidantwoods committed:
Use static analysis plugin for infection

aidantwoods committed:
Support old phpunit, remove lock file

aidantwoods committed:
Dev flag

aidantwoods committed:
Run both commonmark tests

aidantwoods committed:
Remove travis

aidantwoods committed:
Add RenderStack

This is a special configurable that can be used to alter the
Renderable[] AST before output.

aidantwoods committed:
Add initState to tests

aidantwoods committed:
Minor refactor to use type hints

aidantwoods committed:
Return statebearer instead of state

aidantwoods committed:
Restrict autolinks to https

aidantwoods committed:
Add TransformableRenderable capability

aidantwoods committed:
Pass State to render stack closure

aidantwoods committed:
adding method for Container

aidantwoods committed:
Replacing all function to TransformableRenderable

aidantwoods committed:
Complete docblock

aidantwoods committed:
Transformations must return transformable renderables

aidantwoods committed:
Apply correct state

aidantwoods committed:
Off by 1

aidantwoods committed:
Relax mutation tolerance

aidantwoods committed:
Allow substitution of blocks and inlines

aidantwoods committed:
StateBearer must be constructable from State

aidantwoods committed:
Slightly nicer method name

Normally this name might be too vague for an interface, but implementing
this interface is a fairly special use case. We can expect dedicated
types here (so method name unlikely to be an issue).

aidantwoods committed:
Slightly nicer to allow construction from StateBearer

aidantwoods committed:
Update php-cs-fixer

aidantwoods committed:
Add initial documentation

This isn't yet complete, but is a decent start

aidantwoods committed:
Show importing of Parsedown and State in example

aidantwoods committed:
Remove unnecessary cast

aidantwoods committed:
Ignore invalid docblock in tests

aidantwoods committed:
Add additional explaination for the `State` object

Adds a small non-technical introduction to help with intuition, after
the object appears for the first time.

aidantwoods committed:
Test on 8.1

aidantwoods committed:
Error level 1 instead of totallyTyped

Seems this is now the equivalent and sanctioned way to do this.

aidantwoods committed:
Merge pull request #821 from erusev/documentation-v2.0

Documentation 2.0

aidantwoods committed:
Fix title

BenjaminHoegh commented:
@aidantwoods is there a way to manipulate with `TList` without having to rewrite it?

BenjaminHoegh commented:
> Depending on what you're trying to do, could it be made easier if there was a constructor or more getters made available for `TList`?

Just trying to make tasks using the `- [] my task`

BenjaminHoegh commented on docs/Migrating-Extensions-v2.0.md:
````diff
@@ -0,0 +1,222 @@
+
+```php
+$BlockTypes = $State->get(BlockTypes::class)
+    ->addingMarkedLowPrecedence('*', [Abbreviation::class])
````
Missing information about the different ways to add such as `addingMarketHighPrecedence`

BenjaminHoegh reviewed (changes requested):
need some more information about adding new elements
